File: marvs_payrol/www/leave_request.py
import frappe
from datetime import datetime

# =========================================================
# EMPLOYEE INFO
# =========================================================
@frappe.whitelist(allow_guest=True)
def get_employee_info(employee):

    if not employee:
        return {"success": False, "message": "No employee provided"}

    if not frappe.db.exists("PMS-Employee", employee):
        return {"success": False, "message": "Employee not found"}

    emp = frappe.get_doc("PMS-Employee", employee)

    # Build full name safely
    full_name = " ".join(filter(None, [
        getattr(emp, "first_name", ""),
        getattr(emp, "middle_name", ""),
        getattr(emp, "last_name", "")
    ])).strip()

    return {
        "success": True,
        "data": {
            "employee_name": full_name or getattr(emp, "employee_id", emp.name),
            "employee_status": getattr(emp, "status", "") or getattr(emp, "employment_status", ""),
            "employee_id": emp.name
        }
    }

# =========================
# CREATE LEAVE REQUEST
# =========================
@frappe.whitelist(allow_guest=True)
def create_leave(employee, employee_name, leave_type, from_date, to_date, reason):

    try:

        if not employee or not from_date or not to_date:
            return {"success": False, "message": "Missing required fields"}

        # calculate total days
        d1 = datetime.strptime(from_date, "%Y-%m-%d")
        d2 = datetime.strptime(to_date, "%Y-%m-%d")

        total_days = (d2 - d1).days + 1

        if total_days <= 0:
            return {"success": False, "message": "Invalid date range"}

        doc = frappe.get_doc({
            "doctype": "PMS-Leave Request",
            "employee": employee,
            "employee_name": employee_name,
            "leave_type": leave_type,
            "from_date": from_date,
            "to_date": to_date,
            "reason": reason,
            "status": "Pending",
            "total_days": total_days
        })

        doc.insert(ignore_permissions=True)
        frappe.db.commit()

        return {
            "success": True,
            "message": "Leave request submitted successfully"
        }

    except Exception:
        frappe.log_error(frappe.get_traceback(), "Leave Request Error")

        return {
            "success": False,
            "message": frappe.get_traceback()
        }

File: marvs_payrol/www/leave_request.js
function updateDateTime() {

    let now = new Date();

    let options = {
        weekday: "long",
        year: "numeric",
        month: "long",
        day: "numeric",
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit"
    };

    $("#current_datetime").text(
        now.toLocaleString("en-US", options)
    );
}

frappe.ready(() => {

    let current_employee = null;
    
    updateDateTime();
    setInterval(updateDateTime, 1000);

    // =========================
    // LOAD EMPLOYEE INFO
    // =========================
    function load_employee(employee) {

        frappe.call({
            method: "marvs_payrol.www.leave_request.get_employee_info",
            args: { employee: employee },

            callback: function (r) {

                if (r.message && r.message.success) {

                    $("#employee-name").text(r.message.data.employee_name);
                    $("#employee-status").text(r.message.data.employee_status || "-");

                } else {

                    $("#employee-name").text("-");
                    $("#employee-status").text("-");

                }
            }
        });
    }

    // =========================
    // EMPLOYEE CHANGE
    // =========================
    $("#employee").on("change", function () {

        let employee = $(this).val().trim();

        if (!employee) return;

        if (current_employee === employee) return;

        current_employee = employee;

        $("#history-body").html("");
        $("#employee-name").text("-");
        $("#employee-status").text("-");

        load_employee(employee);
    });

    // =========================
    // SUBMIT LEAVE REQUEST
    // =========================
    $("#submit").click(function (e) {

        e.preventDefault();

        let data = {
            employee: $("#employee").val(),
            employee_name: $("#employee-name").text(),
            leave_type: $("#leave_type").val(),
            from_date: $("#from_date").val(),
            to_date: $("#to_date").val(),
            reason: $("#reason").val()
        };

        if (!data.employee || !data.leave_type || !data.from_date || !data.to_date) {
            $("#message").html(`
                <div class="alert alert-danger">
                    Please fill all required fields
                </div>
            `);
            return;
        }

        $("#submit").prop("disabled", true).text("Submitting...");

        frappe.call({
            method: "marvs_payrol.www.leave_request.create_leave",
            args: data,

            callback: function (r) {

                $("#submit").prop("disabled", false).text("Submit Leave Request");

                if (r.message && r.message.success) {

                    $("#message").html(`
                        <div class="alert alert-success">
                            ${r.message.message}
                        </div>
                    `);

                    // RESET FORM
                    $("#employee").val("");
                    $("#employee-name").text("-");
                    $("#employee-status").text("-");
                    $("#leave_type").prop("selectedIndex", 0);
                    $("#from_date").val("");
                    $("#to_date").val("");
                    $("#reason").val("");

                    setTimeout(() => {
                        window.location.href = "/leave-request-table";
                    }, 1000);

                } else {

                    $("#message").html(`
                        <div class="alert alert-danger">
                            ${r.message.message}
                        </div>
                    `);

                }
            }
        });

    });

});